R/decostand.R

```diff
@@ -269,3 +311,323 @@
+  if (is.na(ROPT)){
+    if (verbose){
+      cat("* OptSpace: Guessing an implicit rank.")
+    }
+    r <- min(max(round(.guess_rank(M_E, nnZ.E)), 2), m-1)
+    if (verbose){
+      cat(paste0('* OptSpace: Guessing an implicit rank: Estimated rank : ',r))
+    }
+  } else {
+    r <- round(ROPT)
+    if ((!is.numeric(r)) || (r<1) || (r>m) || (r>n)){
+      stop("* OptSpace: ROPT should be an integer in [1,min(nrow(x),ncol(x))]")
+    }
+  }
+  
+  ## Preprocessing : NITER : maximum number of iterations
+  if ((is.infinite(NITER))||(NITER<=1)||(!is.numeric(NITER))){
+    stop("* OptSpace: invalid NITER number")
+  }
+  NITER <- round(NITER)
+  
+  m0 <- 10000
+  rho <-  eps*n
+  
+  ## Main Computation
+  rescal_param <- sqrt(nnZ.E*r/(norm(M_E,'f')^2))
+  M_E <- M_E*rescal_param
+  
+  # 1. SVD
+  if (verbose){
+    cat("* OptSpace: Step 2: SVD ...")
+  }
+  svdEt <- svd(M_E)
+  X0 <- svdEt$u[,1:r]
+  X0 <- X0[, rev(seq_len(ncol(X0)))]
+  S0 <- diag(rev(svdEt$d[seq_len(r)]))
+  Y0 <- svdEt$v[, seq_len(r)]
+  Y0 <- Y0[, rev(seq_len(ncol(Y0)))]
+  
+  # 3. Initial Guess
+  if (verbose){
+    cat("* OptSpace: Step 3: Initial Guess ...")
+  }
+  X0 <- X0*sqrt(n)
+  Y0 <- Y0*sqrt(m)
+  S0 <- S0/eps
+  
+  # 4. Gradient Descent
+  if (verbose){
+    cat("* OptSpace: Step 4: Gradient Descent ...")
+  }
+  X <- X0
+  Y <- Y0
+  S <- .aux_getoptS(X, Y, M_E, E)
+  # initialize
+  dist <- array(0, c(1, (NITER+1)))
+  dist[1] <- norm((M_E - (X %*% S %*% t(Y)))*E, 'f') / sqrt(nnZ.E)
+  for (i in seq_len(NITER)){
+    # compute the gradient
+    tmpgrad <- .aux_gradF_t(X, Y, S, M_E, E, m0, rho)
+    W <- tmpgrad$W
+    Z <- tmpgrad$Z
+    # line search for the optimum jump length
+    t <- .aux_getoptT(X, W, Y, Z, S, M_E, E, m0, rho)
+    X <- X + t*W;
+    Y <- Y + t*Z;
+    S <- .aux_getoptS(X, Y, M_E, E)
+    # compute the distortion
+    dist[i+1] <- norm(((M_E - X %*% S %*% t(Y))*E),'f') / sqrt(nnZ.E)
+    if (dist[i+1] < TOL){
+      dist <- dist[1:(i+1)]
+      break
+    }
+  }
+  S <- S/rescal_param  
+  # Return Results
+  out <- list()
+  
+  # re-order Optspace may change order during iters
+  index_order <- order(diag(S), decreasing = TRUE)
+  X <- X[, index_order]
+  Y <- Y[, index_order]
+  S <- S[index_order, index_order]
+  out$X <- X
+  out$S <- S
+  out$Y <- Y
+  out$dist <- dist
+  if (verbose){
+    cat('* OptSpace: estimation finished.')
+  }
+
+  # -------------------------------------------
+
+  # This part is not in the Python/Gemelli implementation
+  # but has been added in R to provide more direct access
+  # to the imputed matrix.
+
+  # Reconstruct the matrix
+  M <- X %*% S %*% t(Y)
+
+  # Centering is common operation supporting output visualization
+  # Center cols to 0
+  M <- as.matrix(scale(M, center=TRUE, scale=FALSE))
+  # Center rows to 0
+  M <- as.matrix(t(scale(t(M), center=TRUE, scale=FALSE)))
+
+  # Imputed matrix; add to the output
+  out$M <- M
+
+  # -------------------------------------------
+
+  return(out)
```

Comment:
It is better style now to not use `return` as R automagically calls `return()` on the value of the last line of a function. Just leave this as `out`.